src/node.cc

```diff
@@ -2487,26 +2491,48 @@ struct node_module* get_linked_module(const char* name) {
   return mp;
 }
 
-// DLOpen is process.dlopen(module, filename).
+// DLOpen is process.dlopen(module, filename, flags).
 // Used to load 'module.node' dynamically shared objects.
 //
 // FIXME(bnoordhuis) Not multi-context ready. TBD how to resolve the conflict
 // when two contexts try to load the same shared object. Maybe have a shadow
 // cache that's a plain C list or hash table that's shared across contexts?
 static void DLOpen(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
+  bool is_dlopen_error = false;
+  const char* dlerrmsg = nullptr;
   uv_lib_t lib;
 
   CHECK_EQ(modpending, nullptr);
 
-  if (args.Length() != 2) {
-    env->ThrowError("process.dlopen takes exactly 2 arguments.");
+  if (args.Length() < 2) {
+    env->ThrowError("process.dlopen needs at least 2 arguments.");
     return;
   }
 
+  int32_t flags = RTLD_LAZY;
+  if (args.Length() > 2 && !args[2]->Int32Value(env->context()).To(&flags)) {
+    return env->ThrowTypeError("flag argument must be an integer.");
+  }
+
   Local<Object> module = args[0]->ToObject(env->isolate());  // Cast
   node::Utf8Value filename(env->isolate(), args[1]);  // Cast
-  const bool is_dlopen_error = uv_dlopen(*filename, &lib);
+
+#ifdef __POSIX__
+  // libuv does not support passing flags to dlopen,
+  // so we need to open code the uv_dlopen implementation.
+  lib.errmsg = nullptr;
+  lib.handle = dlopen(*filename, flags);
+  if (!lib.handle) {
+    is_dlopen_error = true;
+    dlerrmsg = dlerror();
+  }
```

Comment:
This still mixes libuv code with direct calls, which I don't think is a good idea.  I'd use libuv on Windows and plain dlopen/dlerror on UNIX.  Maybe split it off into separate functions to avoid any confusion.

If it makes it easier, you can remove the "Module did not self-register" uv_dlclose() call because I don't think that one is sound.  Code from the shared object can have run so it's not safe to unload it again.